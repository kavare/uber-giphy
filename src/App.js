import React, { useState, useEffect, useCallback } from 'react';
import './App.scss';
import logo from './config/logo.svg';
import iconSingleColumn from './config/single-column.svg';
import iconThreeColumns from './config/three-columns.svg';

import {
  NavBar,
  Layout,
  List,
  SearchBar,
  IconButton,
} from './components';

import {
  getSearchResults,
} from './utils';

import {
  useDebounce,
  useInfiniteScroll,
} from './hooks';

function App() {
  const [keywords, setKeywords] = useState('');
  const [isPristine, setIsPristine] = useState(true);
  const [results, isLoading, hasError, setQuery] = useInfiniteScroll(getSearchResults);
  const debouncedKeywords = useDebounce(keywords, 500);
  const search = useCallback(setQuery);
  const [columns, setColumns] = useState(1);

  useEffect(() => {
    if (debouncedKeywords === undefined) return;
    search(debouncedKeywords);
  }, [debouncedKeywords, search]);


  const updateKeywords = (e) => {
    let keywords = e.target.value;
    setKeywords(keywords);

    // [NOTE] UX Improvement: prevent flashing no results screen when users start
    //        searching from initial state
    setTimeout(() => setIsPristine(false), 1000);
  }

  const backToHome = (e) => {
    setKeywords('');
    setIsPristine(true);
  }

  return (
    <div className="ug-app">
      <NavBar>
        <img
          className="ug-navbar__icon"
          src={logo}
          alt="Uber Giphy"
          title="Click to resume to initial state"
          onClick={backToHome}
        />
        <SearchBar
          className="ug-navbar__search"
          keywords={keywords}
          updateKeywords={updateKeywords}
        />
        <div className="ug-navbar__button-group">
          <IconButton
            className={`
              ug-navbar__button
              ${columns === 1 ? 'ug-icon-button--is-active' : ''}
            `}
            img={iconSingleColumn}
            text="Single Column Layout"
            onClick={() => setColumns(1)}
          />
          <IconButton
            className={`
              ug-navbar__button
              ${columns === 3 ? 'ug-icon-button--is-active' : ''}
            `}
            img={iconThreeColumns}
            text="Three Columns Layout"
            onClick={() => setColumns(3)}
          />
        </div>
      </NavBar>
      <Layout>
        { hasError &&
          <h4 className="ug-message ug-message--is-error">
            <span role="img" aria-label="lost connection"> üïµ </span>You might lose the internet connection, please try again
          </h4>
        }
        { isLoading &&
          <h4 className="ug-message ug-message--is-success">
            <span role="img" aria-label="loading"> ‚è≥ </span>Loading...
          </h4>
        }
        <List
          columns={columns}
          items={results}
          isPristine={isPristine}
          isLoading={isLoading}
        />
      </Layout>
    </div>
  );
}

export default App;
