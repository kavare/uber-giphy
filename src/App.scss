@import './config/_variables';

body {
  background-color: $ug-background-color;
}

.ug-app {
  width: 100vw;

  .ug-message {
    position: fixed;
    width: 100%;
    margin: 0;
    padding: $ug-padding / 2;
    font-size: 1.2em;
    font-weight: 200;
    color: $ug-font-color-light;

    &--is-success { background-color: $ug-success-color; }
    &--is-warning { background-color: $ug-warning-color; }
    &--is-error { background-color: $ug-error-color; }
  }
}
