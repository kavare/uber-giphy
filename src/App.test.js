import React from 'react';
import { render, fireEvent, cleanup, wait } from '@testing-library/react';
import App from './App';
import mockData from './config/data.json';

describe('App Integration Test', () => {
  beforeEach(() => {
    fetch.resetMocks();
  })

  afterEach(cleanup);

  test('should matches snapshot', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render with default title with Navbar and List components', () => {
    const { getByText, getByTestId } = render(<App />);
    expect(getByText(/Uber Giphy/i)).toBeInTheDocument();
    expect(getByTestId('ug-navbar')).toBeInTheDocument();
    expect(getByTestId('ug-list')).toBeInTheDocument();
  });

  test('should get response from API and render on the screen after key-in keywords', async () => {
    fetch.mockResponse(JSON.stringify(mockData));
    const { getByTestId, getByText, getByAltText } = render(<App />);
    const searchInput = getByTestId('ug-searchbar__input');
    fireEvent.change(searchInput, {target: {value: 'santa'}});

    await wait(() => {
      expect(fetch.mock.calls.length).toEqual(1)
      expect(getByTestId('ug-list').children.length).toBe(5);
      expect(getByText(mockData.data[0].title)).toBeInTheDocument();
      expect(getByAltText(mockData.data[0].title)).toBeInTheDocument();
    })
  })

  test('should load more items when scroll to bottom', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockData));
    const { getByTestId, getByText, getByAltText } = render(<App />);
    const searchInput = getByTestId('ug-searchbar__input');
    const searchList = getByTestId('ug-list');
    fireEvent.change(searchInput, {target: {value: 'santa'}});

    await wait(() => {
      expect(searchList.children.length).toEqual(5);
      expect(getByText(mockData.data[0].title)).toBeInTheDocument();
      expect(getByAltText(mockData.data[0].title)).toBeInTheDocument();
    })

    const dummyData = Object.assign({}, mockData, {data: [{
      id: 'dummyabc',
      title: 'dummy',
      images: {
        fixed_width: {webp: '', url: ''},
      }
    }]})
    fetch.mockResponseOnce(JSON.stringify(dummyData))

    await fireEvent.scroll(window, {target: {scrollY: 5000}});
    await wait(() => {
      expect(fetch.mock.calls.length).toEqual(2);
      expect(searchList.children.length).toEqual(6);
    })
  })
})