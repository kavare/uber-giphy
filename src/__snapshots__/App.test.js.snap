// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Integration Test should matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ug-app"
  >
    <header
      class="ug-navbar"
      data-testid="ug-navbar"
    >
      <img
        alt="Uber Giphy"
        class="ug-navbar__icon"
        src="logo.svg"
        title="Click to resume to initial state"
      />
      <div
        class="ug-searchbar ug-navbar__search"
        data-testid="ug-searchbar"
      >
        <input
          class="ug-searchbar__input"
          data-testid="ug-searchbar__input"
          placeholder="Search Giphy..."
          type="text"
          value=""
        />
      </div>
      <div
        class="ug-navbar__button-group"
      >
        <img
          alt="Single Column Layout"
          class="ug-icon-button 
              ug-navbar__button
              ug-icon-button--is-active
            "
          data-testid="ug-icon-button"
          src="single-column.svg"
        />
        <img
          alt="Three Columns Layout"
          class="ug-icon-button 
              ug-navbar__button
              
            "
          data-testid="ug-icon-button"
          src="three-columns.svg"
        />
      </div>
    </header>
    <section
      class="ug-layout"
      data-testid="ug-layout"
    >
      <ul
        class="
        ug-list
        ug-list--single-column
        ug-list--no-results
        ug-list--is-pristine
     "
        data-testid="ug-list"
      >
        <li
          class="ug-list__item ug-list__item--is-pristine"
        >
          <h2
            class="ug-list__item-title"
          >
            Uber Giphy
          </h2>
          <h4
            class="ug-list__item-subtitle"
          >
            crafted by kavare
            <span
              aria-label="love"
              role="img"
            >
               ðŸ˜Š 
            </span>
          </h4>
        </li>
      </ul>
    </section>
  </div>
</DocumentFragment>
`;
